Create User Invitation API

Jtang-1 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
Adds the `POST` endpoint `https://www.commcarehq.org/a/[domain]/api/invitation/v1/` to create web user invitation.

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
[USH-5183](https://dimagi.atlassian.net/browse/USH-5183)
[Brief tech spec doc](https://docs.google.com/document/d/1Og-QOdF40B1N3hPrzAmb4-vuyXva6MIpzwbVmjuH150/edit?tab=t.0)

The supported parameters are: email, role, primary_location, assigned_locations, profile, custom_user_data, tableau_role, tableau_groups.

In this PR, I also made some update to the validation in preparation for it needing to support both creation and updates. `None` and an empty string `""` can be treated the same on creation. But when doing an update, caution needs to be done to handle `None` (where no update is done to that parameter) and an empty string `""` (where the parameter is being set to None) differently. On update, most of the time, validation for the parameter can be skipped if it is not updated but validation needs to be done if it is being set to `""`.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
no feature flag

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally. This is a completely new endpoint so does not affecting any existing features. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
There exists tests for API input parameters. This PR adds a test that the endpoint successfully creates and Invitation.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-5183]: https://dimagi.atlassian.net/browse/USH-5183?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Jtang-1 committed:
bug: preoperty does not need to be invoked. Also missing argument

Jtang-1 committed:
bug: correctly handle "None" vs "''" inputs.

There is a slight difference in how bulk upload and the API works. For bulk upload, the existing
user information is typically included. So if the field is required, then the existing value is expected
to be included in the bulk upload. If the field is either empty ("''") or does not exist (None), validation
will failed as "the field is required" for bulk upload. However for the API, if it is a PUT and the field is not
included (None), then that should not fail the required validation.

Jtang-1 committed:
refactor: the validation in LocationValidator is only applicable if location is changed. Separate this out from User Access

Jtang-1 committed:
adds validation for required fields and distinguinsh allowed parameters between post and put

Jtang-1 committed:
require primary and assigned locations to be provided at the same time to be explicit that primary location is one of the assigned locations

Jtang-1 committed:
set up InvitationResource

Jtang-1 committed:
bug: these values should be updated even if there is an existing Invitation

Jtang-1 committed:
feat: create access function for coordinating all validation

Jtang-1 committed:
feat: perform validation upon post

Jtang-1 committed:
feat: create invitation upon post

Jtang-1 committed:
bug: cache only the portion that takes long. The filtering varies per function call so can not be cached by only the domain

Jtang-1 committed:
dehydrate fields to display string representation of objects

Jtang-1 committed:
move InvitationResource to v1

Jtang-1 committed:
Merge branch 'master' into jt/create_user_invitation_api

Jtang-1 committed:
test invitation resource post

Jtang-1 committed:
spacing fix

MartinRiese reviewed (commented):
Looks pretty good to me. But I don't have much django api experience, so I would recommend a second pair of eyes.

Jtang-1 committed:
skip validation of tableau groups and tableau role if there is no input

Jtang-1 committed:
fix test failure for test_simple_is_valid

Jtang-1 committed:
test: set up profile to avoid mocking

Jtang-1 committed:
fix: prevents user data from being set directly if it's already defined within the profile

Jtang-1 committed:
bug: the kwargs are used to match and the default is used to update or create the object

Jtang-1 committed:
for a webuser, its email should be used not its username

Jtang-1 committed:
prefactor: extract out getting location ids from sitecode so that the validation can be reused for API

Jtang-1 committed:
use location_id isn't of location site code as the parameter input. The rationale is because the sitecode value can change

Jtang-1 committed:
bug: needs the location ids not the location object

Jtang-1 committed:
bug: correctly handle if error is a lazytext

Jtang-1 committed:
Merge branch 'master' into jt/create_user_invitation_api

Jtang-1 committed:
test: update to support location id as input instead of sitecode

Jtang-1 committed:
test: update test for call to new validation function

Jtang-1 committed:
test: update parameters for switching to location ids instead of sitecode

Jtang-1 committed:
refactor/bug: parameter names for api and invite won't always align so extract that validation out to be independent

Jtang-1 committed:
define a dataclass to consolidate api parameter inputs

This is to allow the api parameter keys to be decoupled from the validation key names

Jtang-1 committed:
delete print statement

Jtang-1 committed:
bug: use data from spec after it has been popped

Jtang-1 commented:
Been making a bunch of changes... but I should be done now. Bumping for review

Jtang-1 committed:
Merge branch 'master' into jt/create_user_invitation_api

Robert-Costello commented on corehq/apps/api/tests/test_validation.py:
```diff
@@ -103,15 +118,9 @@ def test_validate_role_with_no_role_input(self):
         self.definition.profile_required_for_user_type = [UserFieldsView.WEB_USER]
         self.definition.save()
-        with patch.object(WebUserResourceValidator,
-                          'profiles_by_name', new_callable=PropertyMock) as mock_profiles_by_name:
```
Just out of curiosity, was there an issue with mocking or just a preference to not use it?

Jtang-1 replied:
mocking also worked. But the non-mocking route required minimal set up so I preferred that route since it's a truer test.

Robert-Costello commented on corehq/apps/api/validation.py:
```diff
@@ -100,6 +104,21 @@ def validate_custom_data(self, custom_data, profile_name):
+
+        for key in custom_data.keys():
+            if key in system_fields:
+                errors.append(_("'{}' cannot be set directly").format(key))
```
Not blocking, but would it be helpful if this message was a little more descriptive or instructive? By that I mean pointing out that this field is part of a profile and the individual fields can't be updated and the user would need to update the profile.

Jtang-1 replied:
[09d6a55](https://github.com/dimagi/commcare-hq/pull/35575/commits/09d6a55393934dbf26b6bc728ca720dac7c99725)

esoergel commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+        if spec.assigned_location_ids:
+            primary_loc = SQLLocation.active_objects.get(
+                location_id=spec.primary_location_id)
+            assigned_locs = SQLLocation.active_objects.filter(
+                location_id__in=spec.assigned_location_ids, domain=domain)
+            real_ids = [loc.location_id for loc in assigned_locs]
+
+            if missing_ids := set(spec.assigned_location_ids) - set(real_ids):
+                raise ImmediateHttpResponse(JsonResponse(
```
Could the location validation stuff go in the validator class?

Jtang-1 replied:
Yes it can, I decided to put it here to avoid having to fetch the location objects again. But maybe that extra fetch isn't a big deal, would definitely be cleaner to do the validation inside the validator class

esoergel replied:
Ah, I definitely do support minimizing lookups, especially for an API, which can easily get spammed heavily.  The validator code already looks up locations, though I think it's pretty far deep - not sure how straightforward it'd be to get access to it - it's inside `validate_assigned_locations_allow_users`

Robert-Costello reviewed (approved):
Left a couple of comments. Looks good to me otherwise.

esoergel commented on corehq/apps/api/validation.py:
```diff
@@ -11,16 +20,64 @@
+from corehq.toggles import TABLEAU_USER_SYNCING
+
+
+@dataclass
```
This `dataclass` plus custom validator class makes me think of `attrs`, which has a [built-in system for writing custom validation](https://www.attrs.org/en/stable/init.html#validators).  That'd I think cut down on some of the boilerplate, and reduce the mental burden of understanding a new system.

I do also see that tastypie lets you [pass in a Django form](https://django-tastypie.readthedocs.io/en/latest/validation.html#cleaneddataformvalidation) for validation.  That'd be cool if we could reuse the same form class for the UI, the API, and the importer, but I'm guessing there'd be some blockers.

Jtang-1 replied:
I'll look into `attrs`

It would be nice to use the same django form for validation, but would require some overhauling. I already did some refactoring to consolidate the validation between the UI and importer in a [previous PR](https://github.com/dimagi/commcare-hq/pull/35448) and tried in this PR to reuse importer validation logic for the API.

Jtang-1 replied:
Thanks for that suggestion - it overall is a bit nicer. I implemented it in the follow up "Update Web User PR" to avoid it conflicting with the files in this PR https://github.com/dimagi/commcare-hq/pull/35617/commits/25b7fe21458b9bf4b3b299e0fcfb722e2ea8da1f

Jtang-1 committed:
clarify error message for conflicting user data and profile

Jtang-1 commented on corehq/apps/api/validation.py:
```diff
@@ -11,16 +20,64 @@
+
+
+@dataclass
+class WebUserSpec:
```
I'm naming it `WebUser` because there's a followup PR for "updating web user" and I'm using the same validation classes. Given that, what do you think about the naming? 

I was influenced by how it's handled in bulk upload, we seem to treat Invitations as just another web user when validating

esoergel replied:
Nice, that sounds great - I would definitely like to see more parity between the two (even _eventually_ the same DB model once this is in SQL?)

Jtang-1 commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+            custom_user_data=bundle.data.pop('custom_user_data', None),
+            tableau_role=bundle.data.pop('tableau_role', None),
+            tableau_groups=bundle.data.pop('tableau_groups', None),
+            unhandled_data=bundle.data,
```
No, tastypie surprisingly doesn't fail. It just ignores the extra fields. This would have been fine, but I thought it would be nicer to error

esoergel replied:
that's frustrating

Jtang-1 commented on corehq/apps/api/validation.py:
```diff
@@ -11,16 +20,64 @@
+        spec_dict = {k: v for k, v in asdict(self).items()
+                     if k != 'unhandled_data' and v is not None}
+        if self.unhandled_data:
+            spec_dict.update(self.unhandled_data)
```
the `unhandled_data` field as solely introduced to be used for validating the parameters. I needed unhandled data to be included when getting the flat dict so that `validate_parameters` can check the keys. 

But I simplified this logic in the followup "Update Web User API" PR - see this commit from that PR https://github.com/dimagi/commcare-hq/pull/35617/commits/c93e6bc2e27e7a1e3a7049c2b3a0a1a1fa95edc6

esoergel replied:
ah gotcha - okay, happy to table this conversation to that PR

Jtang-1 committed:
remove fluff

1. avoid pulling whole objects from DB and fetch just location ids
2. primary_location location id can be accessed directly via primary_location_id
3. change "errors" to "error" since "errors" is used elsewhere as a list

Jtang-1 commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+    def obj_create(self, bundle, **kwargs):
+        domain = kwargs['domain']
+        validator = WebUserResourceValidator(domain, bundle.request.couch_user)
+        spec = WebUserSpec(
```
It makes sense to use the unpacking operator here since the `WebUserSpec` key matches the TastyPie field names. But I purposely introduced the `WebUserSpec` to decouple the key names used for validation and the TastyPie field names. This came in handy in the [followup PR](https://github.com/dimagi/commcare-hq/pull/35617) [here](https://github.com/dimagi/commcare-hq/blob/61357f39d47e1b1cd3bf384e0298de689068d880/corehq/apps/api/resources/v1_0.py#L104-L116)

Although I still need to directly reference the Tastypie field names when doing `validate_parameters`...

Jtang-1 committed:
avoid unnecessary fetch of primary location

Jtang-1 commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+        assigned_locs = []
+        if spec.assigned_location_ids:
+            primary_loc = SQLLocation.active_objects.get(
+                location_id=spec.primary_location_id)
```
ahh, I did not know I could do that  [c91c1e8](https://github.com/dimagi/commcare-hq/pull/35575/commits/c91c1e8d2ef66b879d6ffdbb46b1e01fad6fa698)

Jtang-1 commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+
+            if missing_ids := set(spec.assigned_location_ids) - set(real_ids):
+                raise ImmediateHttpResponse(JsonResponse(
+                    {"errors": f"Could not find location ids: {', '.join(missing_ids)}."}, status=400))
```
fixed here [8927a13](https://github.com/dimagi/commcare-hq/pull/35575/commits/8927a13406ef89f507c6b27b583acf6a465e0cb6)

Jtang-1 commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+
+    def dehydrate_primary_location_id(self, bundle):
+        if bundle.obj.primary_location:
+            return bundle.obj.primary_location.location_id
```
I like that

Jtang-1 commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+        return bundle.obj.get_role_name()
+
+    def dehydrate_assigned_location_ids(self, bundle):
+        return [loc.location_id for loc in bundle.obj.assigned_locations.all() if loc is not None]
```
[8927a13](https://github.com/dimagi/commcare-hq/pull/35575/commits/8927a13406ef89f507c6b27b583acf6a465e0cb6)

Jtang-1 committed:
test: update assertion string

Jtang-1 commented on corehq/apps/api/resources/v1_0.py:
```diff
@@ -49,3 +64,90 @@ def prepend_urls(self):
+            return bundle.obj.primary_location.location_id
+
+    def dehydrate_tableau_groups(self, bundle):
+        return [group.name for group in get_tableau_groups_by_ids(bundle.obj.tableau_group_ids,
+                                                                 bundle.request.domain)]
```
It was considered to support list endpoint as part of this work but it was dropped. I don't think there is any immediate plan to support list endpoint. 

Regarding `get_tableau_groups_by_ids`, the performance intensive portion of that function is `get_tableau_group_json` (which requests data from Tableau). I cached `get_tableau_group_json` with `@quickcache(['domain'], timeout=2 * 60)` so I think that should be adequate. What do you think?

esoergel replied:
Yeah that should be fine.

Jtang-1 committed:
delete file that was accidently added

Jtang-1 commented:
Bumping for rereview/ approval if things look good

Robert-Costello reviewed (approved):
Looks like major concerns have been addressed

Jtang-1 commented:
Thank you for the reviews!
